src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.Builder.StatesWrapper.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis;
+
+internal sealed partial class NodeStateTable<T>
+{
+    public sealed partial class Builder
+    {
+        /// <summary>
+        /// Helper type that keeps track of the TableEntrys that are being added and will reuse the same entries from
+        /// the previous node table if we're producing the exact same ones.
+        /// </summary>
+        private struct StatesWrapper
+        {
+            private readonly NodeStateTable<T> _previous;
+            private readonly IEqualityComparer<T> _equalityComparer;
+
+            /// <summary>
+            /// Note: we could potentially lazily initialize this the first time we see ourselves deviating from
+            /// _previous.  However, this approach keeps things very simple here.
+            /// </summary>
+            private readonly ArrayBuilder<TableEntry> _tableEntries;
+
+            /// <summary>
+            /// Tracks if the new table we're building has all the same entries as the previous table.  If we end up
+            /// with the same set of entries at the end, we can just point our new table at that same array, avoiding a
+            /// costly allocation.
+            /// </summary>
+            private bool _unchangedFromPrevious = true;
```

Comment:
>_unchangedFromPrevious [](http://example.com/codeflow?start=25&length=22)

Is there a reason to track this value, or could we simply compare `_tableEntries` with `_previous._states` in `ToImmutableAndFree()`? And if we can we compare in `ToImmutableAndFree()`, could we do so in the original `Builder` instance, without adding `StatesWrapper`?

Reply:
i originally wanted to avoid a linear scan over the data.  so i was doing it just as the data streamed in.  but that doesn't really seem that relevant or useful, so i'll switch to your approach :)

Reply:
Done!